From cdd3cb7b5040162c82d8e2b6b9e37f4bda89a18d Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Thu, 16 Mar 2023 19:32:00 +0530
Subject: [PATCH 118/500] OpenWrt: 260-crypto_test_dependencies.patch

kernel: prevent cryptomgr from pulling in useless extra dependencies for tests that are not run

Reduces kernel size after LZMA by about 5k on MIPS

lede-commit: 044c316167e076479a344c59905e5b435b84a77f
Signed-off-by: Felix Fietkau <nbd@nbd.name>

Change-Id: I54d14e261bcea9acd13388463bd125c44bf3ece8
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
(cherry picked from commit cd83d75ef9e113029ae462184a45526cab6d2b8d)
---
 crypto/Kconfig   | 18 +++++++++---------
 crypto/algboss.c |  4 ++++
 2 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/crypto/Kconfig b/crypto/Kconfig
index b9c68e4daf7b..6433d521b57e 100644
--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -148,15 +148,15 @@ config CRYPTO_MANAGER
 	  cbc(aes).
 
 config CRYPTO_MANAGER2
-	def_tristate CRYPTO_MANAGER || (CRYPTO_MANAGER!=n && CRYPTO_ALGAPI=y)
-	select CRYPTO_ACOMP2
-	select CRYPTO_AEAD2
-	select CRYPTO_AKCIPHER2
-	select CRYPTO_SIG2
-	select CRYPTO_HASH2
-	select CRYPTO_KPP2
-	select CRYPTO_RNG2
-	select CRYPTO_SKCIPHER2
+	def_tristate CRYPTO_MANAGER || (CRYPTO_MANAGER!=n && CRYPTO_ALGAPI=y && !CRYPTO_MANAGER_DISABLE_TESTS)
+	select CRYPTO_ACOMP2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_AEAD2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_AKCIPHER2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_SIG2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_HASH2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_KPP2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_RNG2 if !CRYPTO_MANAGER_DISABLE_TESTS
+	select CRYPTO_SKCIPHER2 if !CRYPTO_MANAGER_DISABLE_TESTS
 
 config CRYPTO_USER
 	tristate "Userspace cryptographic algorithm configuration"
diff --git a/crypto/algboss.c b/crypto/algboss.c
index 0de1e6697949..333e601169f4 100644
--- a/crypto/algboss.c
+++ b/crypto/algboss.c
@@ -204,6 +204,10 @@ static int cryptomgr_schedule_test(struct crypto_alg *alg)
 	memcpy(param->alg, alg->cra_name, sizeof(param->alg));
 	param->type = alg->cra_flags;
 
+#ifdef CONFIG_CRYPTO_MANAGER_DISABLE_TESTS
+	param->type |= CRYPTO_ALG_TESTED;
+#endif
+
 	thread = kthread_run(cryptomgr_test, param, "cryptomgr_test");
 	if (IS_ERR(thread))
 		goto err_free_param;
-- 
2.34.1

